src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2030,296 +1406,42 @@ ObjectMonitor* ObjectSynchronizer::inflate(Thread* self, oop object,
-// an async deflation is in progress. There are a number of checks
-// as part of the protocol to make sure that the calling thread has
-// not lost the race to a contending thread.
-//
-// The ObjectMonitor has been successfully async deflated when:
-//   (contentions < 0)
-// Contending threads that see that condition know to retry their operation.
-//
-bool ObjectSynchronizer::deflate_monitor(ObjectMonitor* mid,
-                                         ObjectMonitor** free_head_p,
-                                         ObjectMonitor** free_tail_p) {
-  // A newly allocated ObjectMonitor should not be seen here so we
-  // avoid an endless inflate/deflate cycle.
-  assert(mid->is_old(), "must be old: allocation_state=%d",
-         (int) mid->allocation_state());
-
-  if (mid->is_busy()) {
-    // Easy checks are first - the ObjectMonitor is busy so no deflation.
-    return false;
-  }
-
-  const oop obj = mid->object_peek();
-
-  if (obj == NULL) {
-    // If the object died, we can recycle the monitor without racing with
-    // Java threads. The GC already broke the association with the object.
-    mid->set_owner_from(NULL, DEFLATER_MARKER);
-    mid->_contentions = -max_jint;
-  } else {
-    // Set a NULL owner to DEFLATER_MARKER to force any contending thread
-    // through the slow path. This is just the first part of the async
-    // deflation dance.
-    if (mid->try_set_owner_from(NULL, DEFLATER_MARKER) != NULL) {
-      // The owner field is no longer NULL so we lost the race since the
-      // ObjectMonitor is now busy.
-      return false;
+  while (iter.has_next()) {
+    if (deflated_count >= (size_t)MonitorDeflationMax) {
+      break;
     }
-
-    if (mid->contentions() > 0 || mid->_waiters != 0) {
-      // Another thread has raced to enter the ObjectMonitor after
-      // mid->is_busy() above or has already entered and waited on
-      // it which makes it busy so no deflation. Restore owner to
-      // NULL if it is still DEFLATER_MARKER.
-      if (mid->try_set_owner_from(DEFLATER_MARKER, NULL) != DEFLATER_MARKER) {
-        // Deferred decrement for the JT EnterI() that cancelled the async deflation.
-        mid->add_to_contentions(-1);
-      }
-      return false;
+    ObjectMonitor* mid = iter.next();
+    if (mid->deflate_monitor()) {
+      deflated_count++;
     }
 
-    // Make a zero contentions field negative to force any contending threads
-    // to retry. This is the second part of the async deflation dance.
-    if (Atomic::cmpxchg(&mid->_contentions, (jint)0, -max_jint) != 0) {
-      // Contentions was no longer 0 so we lost the race since the
-      // ObjectMonitor is now busy. Restore owner to NULL if it is
-      // still DEFLATER_MARKER:
-      if (mid->try_set_owner_from(DEFLATER_MARKER, NULL) != DEFLATER_MARKER) {
-        // Deferred decrement for the JT EnterI() that cancelled the async deflation.
-        mid->add_to_contentions(-1);
+    if (self->is_Java_thread() &&
```

Comment:
Again unclear how a non-JavaThread is doing this? Isn't this always done by the MonitorDeflation thread??

Reply:
This function, ObjectSynchronizer::deflate_monitor_list(), may be called
by either the MonitorDeflationThread or the VMThread. The VMThread
does not need to block for the safepoint which is what the
if (self->is_Java_thread() prevents.